pkg/comp-functions/functions/spksmariadb/pvcresize.go

```diff
@@ -0,0 +1,260 @@
+
+		sts := &appsv1.StatefulSet{}
+		err = svc.GetObservedKubeObject(sts, comp.Name+"-sts-observer")
+		if err != nil && err != runtime.ErrNotFound {
+			return runtime.NewFatalResult(fmt.Errorf("cannot get observed statefulset job: %w", err))
+		}
+
+		// If the xkube object has been created it's still possible that the actual job hasn't been observedJob.
+		observedJob := len(xJob.Status.AtProvider.Manifest.Raw) > 0
+
+		// Check the sts if it has been updated
+		stsSize := int64(0)
+		if len(sts.Spec.VolumeClaimTemplates) > 0 {
+			stsSize, _ = sts.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage().AsInt64()
+		}
+		newSize, found, err := unstructured.NestedString(values, "persistence", "size")
+		if !found {
+			return runtime.NewFatalResult(fmt.Errorf("disk size not found in observed release"))
+		}
+
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("failed to read size from observed release: %w", err))
+		}
+		desiredSize, err := getSizeAsInt(newSize)
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("cannot parse desired size: %w", err))
+		}
+		stsUpdated := stsSize == desiredSize
+
+		deletionJob := &batchv1.Job{}
+		if observedJob {
+			err := json.Unmarshal(xJob.Status.AtProvider.Manifest.Raw, deletionJob)
+			if err != nil {
+				return runtime.NewFatalResult(fmt.Errorf("cannot unmarshal sts deleter job: %w", err))
+			}
+		}
+
+		// The job hasn't been observed yet, so we need to keep it in desired, or we will have a recreate loop
+		// Also as long as it hasn't finished we need to make sure it exists.
+		if (!observedJob || deletionJob.Status.Succeeded < 1) || (sts.Status.ReadyReplicas == 0 && !stsUpdated) {
+			err := addDeletionJob(svc, comp, newSize, release.GetName())
+			if err != nil {
+				return runtime.NewFatalResult(fmt.Errorf("cannot create RBAC for the deletion job: %w", err))
+			}
+		}
+		return nil
+	}
+
+	err = addStsObserver(svc, comp)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Errorf("cannot observe sts: %w", err).Error())
+	}
+
+	sts := &appsv1.StatefulSet{}
+	err = svc.GetObservedKubeObject(sts, comp.Name+"-sts-observer")
+	if err == runtime.ErrNotFound {
+		return nil
+	}
+	if err != nil {
+		return runtime.NewFatalResult(fmt.Errorf("cannot get observed statefulset: %w", err))
+	}
+	stsSize := int64(0)
+	// Check the current size in the sts
+	if len(sts.Spec.VolumeClaimTemplates) > 0 {
+		stsSize, _ = sts.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage().AsInt64()
+	}
+
+	patch, newSize, result := needReleasePatch(values, stsSize)
+	if result != nil {
+		return result
+	}
+
+	if patch {
+		err = addDeletionJob(svc, comp, newSize, release.GetName())
+		if err != nil {
+			return runtime.NewFatalResult(fmt.Errorf("cannot create the deletion job: %w", err))
+		}
+		// We pause the release at this point to make sure that provider-helm doesn't update the
+		// release until the deletion job removed the sts
+		release.SetAnnotations(map[string]string{
```

Comment:
I'd probably set the annotation via the script as well. There's still potential for a very narrow race condition here.

Reply:
I think that setting it in script has exactly the same risk of that race condition. Here it's maybe even smaller as annotation will be set immediately, even before job starts.

Reply:
I wouldn't bet on that, we don't know when crossplane decides to actually reconcile this. It could be that it gets reconciled after the job has spawned. 

That's why I would move it to the script. `kubectl` will return after it sent the request directly to the API Server, which will greatly reduce the risk of this race condition happening.

Reply:
As discussed, moving it to the script is not really an option, as both the script and the release will be reconciled at the same time.
I adjusted the script though to check and wait for the annotation on the release.